File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {

public:
void marker(TreeNode* root, unordered_map<TreeNode*, TreeNode*>&parent)
{
    queue<TreeNode*>q;

    q.push(root);
    while(!q.empty())
    {
        TreeNode*front=q.front();
        q.pop();

        if(front->left)
        {
            parent[front->left]=front;
            q.push(front->left);
        }
        if(front->right)
        {
            parent[front->right]=front;
            q.push(front->right);
        }
    }

    
}
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {

        unordered_map<TreeNode*, TreeNode*>parent;

        marker(root, parent);

        queue<TreeNode*>q;
        q.push(target);

        unordered_map<TreeNode*,bool>visited;
        visited[target]=true;

        int curr=0;

        while(!q.empty())
        {
            int size=q.size();
           
            if(curr==k) break;
            curr++;
            for(int i=0;i<size;i++)
            {
                TreeNode* front=q.front();
                q.pop();
                


                if(front->left && visited[front->left]!=true)
                {
                    q.push(front->left);
                    visited[front->left]=true;

                }
                if(front->right && visited[front->right]!=true)
                {
                    q.push(front->right);
                    visited[front->right]=true;

                }
                if(parent[front] && !visited[parent[front]])
                {
                    q.push(parent[front]);
                    visited[parent[front]]=true;

                }
            }
        }
        vector<int>ans;

        while(!q.empty())
        {

            ans.push_back(q.front()->val);
            q.pop();

        }


        return ans;


        

        
    }
};

File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        adj=defaultdict(list)

        def dfs(parent, node):
            if not node:
                return
            if not parent:
                if node.left:
                    adj[node].append(node.left)
                if node.right:
                    adj[node].append(node.right)
            else:
                adj[node].append(parent)
                if node.left:
                    adj[node].append(node.left)
                if node.right:
                    adj[node].append(node.right)
            dfs(node,node.left)
            dfs(node,node.right)





        dfs(None, root)
        visited=set()
        ans=[]
        

        def dfsnei(node,distance):


            if distance==k:
                ans.append(node.val)
            visited.add(node)

            for nei in adj[node]:
                if nei not in visited:
                    visited.add(nei)
                    
                    dfsnei(nei,distance+1)
            #visited.remove(node)

                
            

        dfsnei(target,0)
        return ans

        